BParkingNano/plugins/BtoXLLTemplateFitter.h

```diff
@@ -0,0 +1,241 @@
+                  const std::vector< pat::CompositeCandidate >& vtracks, 
+                  const reco::BeamSpot& vBspot,
+                  const edm::ParameterSet& runParameters,
+                  edm::ESHandle< MagneticField >& vbField ):
+   leptons( vleptons ), tracks( vtracks ), Bspot( vBspot ), bField( vbField )
+{
+
+ // get parameters
+ ptLep1Cut_   = runParameters.getParameter<double>( "ptLep1Cut" );
+ bdtEl1Cut_   = runParameters.getParameter<double>( "bdtEl1Cut" );
+ dzLepLepCut_ = runParameters.getParameter<double>( "dzLepLepCut" );
+ mllMin_      = runParameters.getParameter<double>( "mllMin" );
+ mllMax_      = runParameters.getParameter<double>( "mllMax" );
+ MBMin_       = runParameters.getParameter<double>( "MBMin" );
+ MBMax_       = runParameters.getParameter<double>( "MBMax" );
+ drTrkLepCut_ = runParameters.getParameter<double>( "drTrkLepCut" );
+ BPtCut_      = runParameters.getParameter<double>( "BPtCut" );
+ BVtxProbCut_ = runParameters.getParameter<double>( "BVtxProbCut" );
+ BCosACut_    = runParameters.getParameter<double>( "BCosACut" );
+
+}
+
+template < typename T >
+std::unique_ptr< std::vector<pat::CompositeCandidate> >
+BtoXLLTemplateFitter< T >::ReconstructB( float trackMass )
+{
+  // first we make everything transient to avoid invoking the builder many times
+
+  // templated class; creates transient objects and applies cuts
+  PreFitter prefit(bField);
+
+  // pair creation returns transient tracks cuts ( ptL1 bdt mllmin mllmax dzll)
+  auto tpairs = prefit.CreateTransientPairs( leptons, ptLep1Cut_, bdtEl1Cut_,
+                                             mllMin_, mllMax_, dzLepLepCut_ );
+
+  // transient tracks from composite part
+  auto ttracks = prefit.CreateTransientObject( tracks );
+
+  // provides also indices of good lep pairs
+  auto indices = prefit.GetIndexPairs();
+
+  // output
+  auto cands = std::make_unique< std::vector<pat::CompositeCandidate> >();
+
+  // loop on the pairs of indices of leptons
+  for ( unsigned int ilep = 0; ilep < indices.size(); ++ilep ){
+    unsigned int ilep1( indices[ilep].first );
+    unsigned int ilep2( indices[ilep].second );
+
+    // get 4 vector
+    ROOT::Math::LorentzVector lep1( (leptons[ilep1]).p4() );
+    ROOT::Math::LorentzVector lep2( (leptons[ilep2]).p4() );
+
+    // now loop over tracks and reconstruct  B
+    for ( const pat::CompositeCandidate & trk : tracks ){
+      
+      // get tracks 4 vector
+      ROOT::Math::LorentzVector Kaon( trk.p4() );
+
+      //change mass according to E^2new= E^2old - m^2old + m^2new
+      Kaon.SetE( TMath::Sqrt( pow(Kaon.E(),2) + pow(trackMass,2) 
+                            - pow(Kaon.M(),2) ) );
+
+    
+      // cut on candidate mass and pT
+      if ( ( Kaon + lep1 + lep2 ).M() < MBMin_ ||
+           ( Kaon + lep1 + lep2 ).M() > MBMax_ ||
+           ( Kaon + lep1 + lep2 ).Pt() < BPtCut_ )
+           continue;
+    
+      // dr cut (trk,lep) only on same charge objects
```

Comment:
because what we want to remove here is the possibility a track to come from the lepton. So making a simple Dr plot we saw that this is prominent when they have the same charge (at least for muons). But i can change it, it is really small number of B anyway